python/lsst/daf/butler/direct_query_driver/_driver.py

```diff
@@ -0,0 +1,1200 @@
+                case _:
+                    raise NotImplementedError("TODO")
+        except:  # noqa: E722
+            self._context.__exit__(*sys.exc_info())
+            raise
+
+    def _convert_dimension_record_results(
+        self,
+        raw_rows: Iterable[sqlalchemy.Row],
+        next_key: PageKey | None,
+    ) -> DimensionRecordResultPage:
+        """Convert a raw SQL result iterable into a page of `DimensionRecord`
+        query results.
+
+        Parameters
+        ----------
+        raw_rows : `~collections.abc.Iterable` [ `sqlalchemy.Row` ]
+            Iterable of SQLAlchemy rows, with `Postprocessing` filters already
+            applied.
+        next_key : `PageKey` or `None`
+            Key for the next page to add into the returned page object.
+
+        Returns
+        -------
+        result_page : `DimensionRecordResultPage`
+            Page object that holds a `DimensionRecord` container.
+        """
+        result_spec = cast(DimensionRecordResultSpec, self._result_spec)
+        record_set = DimensionRecordSet(result_spec.element)
+        record_cls = result_spec.element.RecordClass
+        if isinstance(result_spec.element, SkyPixDimension):
+            pixelization = result_spec.element.pixelization
+            id_qualified_name = qt.ColumnSet.get_qualified_name(result_spec.element.name, None)
+            for raw_row in raw_rows:
+                pixel_id = raw_row._mapping[id_qualified_name]
+                record_set.add(record_cls(id=pixel_id, region=pixelization.pixel(pixel_id)))
+        else:
+            # Mapping from DimensionRecord attribute name to qualified column
+            # name, but as a list of tuples since we'd just iterate over items
+            # anyway.
+            column_map = list(
+                zip(
+                    result_spec.element.schema.dimensions.names,
+                    result_spec.element.dimensions.names,
+                )
+            )
+            for field in result_spec.element.schema.remainder.names:
+                if field != "timespan":
+                    column_map.append(
+                        (field, qt.ColumnSet.get_qualified_name(result_spec.element.name, field))
+                    )
+            if not result_spec.element.temporal:
+                for raw_row in raw_rows:
+                    m = raw_row._mapping
+                    record_set.add(record_cls(**{k: m[v] for k, v in column_map}))
+            else:
+                timespan_qualified_name = qt.ColumnSet.get_qualified_name(
+                    result_spec.element.name, "timespan"
+                )
+                for raw_row in raw_rows:
+                    m = raw_row._mapping
+                    d = {k: m[v] for k, v in column_map}
+                    d["timespan"] = self._timespan_repr_cls.extract(m, name=timespan_qualified_name)
+                    record_set.add(record_cls(**d))
```

Comment:
It might be better not to duplicate the logic for the non-timespan columns between these two branches... creates the potential for them to get out of sync as things change.
```suggestion
            if result_spec.element.temporal:
                timespan_qualified_name = qt.ColumnSet.get_qualified_name(
                    result_spec.element.name, "timespan"
                )
            else:
                timespan_qualified_name = None
                
            for raw_row in raw_rows:
                m = raw_row._mapping
                d = {k: m[v] for k, v in column_map}
                if timespan_qualified_name is not None:
                    d["timespan"] = self._timespan_repr_cls.extract(m, name=timespan_qualified_name)
                record_set.add(record_cls(**d))
```

Reply:
Yeah, this was intentional - but possibly premature - optimization; I wanted to bring as much out of the loop over result rows as possible. But your version is all that different in that respect.